sqs/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/SqsSink.scala

```diff
@@ -0,0 +1,335 @@
+    */
+  private def getNextBackoff(lastBackoff: Long): Long =
+    (minBackoff + randomGenerator.nextDouble() * (lastBackoff * 3 - minBackoff)).toLong.min(maxBackoff)
+
+  def shutdown(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10000, MILLISECONDS)
+    ()
+  }
+
+  case class BatchResultErrorInfo private (code: String, message: String)
+}
+
+/**
+  *  SqsSink companion object with factory method
+  */
+object SqsSink {
+
+  /**
+    * Create an SqsSink and schedule a task to flush its EventBuffer.
+    * Exists so that no threads can get a reference to the SqsSink
+    * during its construction.
+    */
+  def createAndInitialize(
+    sqsConfig: Sqs,
+    bufferConfig: BufferConfig,
+    queueName: String,
+    executorService: ScheduledExecutorService
+  ): Either[Throwable, SqsSink] = {
+    val client = for {
+      provider <- getProvider(sqsConfig.aws)
+      client   <- createSqsClient(provider, sqsConfig.region)
+      _ = sqsQueueExists(client, queueName)
+    } yield client
+
+    client.map { c =>
+      val sqsSink = new SqsSink(c, sqsConfig, bufferConfig, queueName, executorService)
+      sqsSink.EventBuffer.scheduleFlush()
+
+      // When the application is shut down try to send all buffered events.
+      Runtime
+        .getRuntime
+        .addShutdownHook(new Thread {
+          override def run(): Unit = {
+            sqsSink.EventBuffer.flush()
+            sqsSink.shutdown()
+          }
+        })
+      sqsSink
+    }
+  }
+
+  private def createSqsClient(provider: AWSCredentialsProvider, region: String) =
+    Either.catchNonFatal(
+      AmazonSQSClientBuilder
+        .standard()
+        .withRegion(region)
+        .withCredentials(provider)
+        .withClientConfiguration(
+          new ClientConfiguration().withRetryPolicy(
+            new RetryPolicy(
+              new RetryCondition {
+                override def shouldRetry(
+                  originalRequest: AmazonWebServiceRequest,
+                  exception: AmazonClientException,
+                  retriesAttempted: Int
+                ): Boolean =
+                  retriesAttempted < 30 //max allowed by SQS client
+              },
+              new PredefinedBackoffStrategies.FullJitterBackoffStrategy(1000, 5 * 3600),
+              30,
+              true,
+              true
+            )
+          )
+        )
+        .build
+    )
+
+  /**
+    * Check whether an SQS queue exists
+    * @param name Name of the queue
+    * @return Whether the queue exists
+    */
+  private def sqsQueueExists(client: AmazonSQS, name: String): Unit = {
+    lazy val log = LoggerFactory.getLogger(getClass())
+    try {
+      client.getQueueUrl(name)
+      ()
+    } catch {
+      case _: QueueDoesNotExistException => log.error(s"SQS queue $name does not exist.")
+      case t: Throwable                  => log.error(t.getMessage)
```

Comment:
Don't we want to crash the app when the queue doesn't exist ?

Reply:
No, we removed that behaviour previously because in case of outage, you can keep providing more resources and keep events in memory, in the hope that it will get fixed before the collector dies.

Reply:
In this scenario the collector is able to receive and buffer events ? Will the initialization be retried until it finally works and send the buffered events ?

Reply:
The initialization does not get retried. The client is initialized (as in, the flush task is scheduled to be executed by it) regardless of whether the destination exists or not. I we try to send an event to an non-existent queue, then it's the job of `sendBatch` to keep retrying. If the queue is then created, then the retry will succeed.

Reply:
I missed the fact that SQS client can be initialized successfully even in case of outage. Great !